Adaptive Hole Map

ackirby88 opened the pull request:
Initial implementation of the adaptive hole map algorithm.
To enable/disable this functionality, set line 64 in _codetypes.h_
```
#define USE_ADAPTIVE_HOLEMAP 1 // [0] original hole map, [1] adaptive hole map
```

I've done preliminary testing on the tioga test case: unstructured tetrahedral sphere and Cartesian background mesh.

**Test Case Input (case/grid/input):**
```
 $inputs
 nlayers=40,
 wallSpacing=1e-5,
 outerSpacing=0.1,
 outerB=0.05,
 xlo(1)=-5.,
 xlo(2)=-5.,
 xlo(3)=-5.,
 xhi(1)=5.,
 xhi(2)=5.,
 xhi(3)=5.,
 dx(1)=0.1,
 dx(2)=0.1,
 dx(3)=0.1,
 $end
```
 
**Original Hole Map** (./run.sh 12):
```
 Generating           24  partitions
 Finished generating grid ..
 Writing tecplot compatible files..
 All done
 # tioga test on           12  processes
 # tioga test : finished reading grids
 connectivity time=  0.71787999999999996     
           2          14
           0           0
           0           0
           1           7
           0           0
           0           0
           2          14
           0           0
           0           0
           3          21
           0           0
           0           0
           5          35
           0           0
           0           0
           0           0
           0           0
           2          14
           0           0
           0           0
           0           0
           0           0
        1202       10818
        1282       11538
 data update time=   7.0409999999998529E-003
-- Interpolation error statistics --
ProcId    Error
   0     0.6619866E-17
   3     0.5463998E-17
   4     0.6485183E-17
   5     0.4214754E-17
   6     0.6526902E-17
   7     0.3255305E-16
   8     0.7567119E-17
   9     0.8691998E-17
  10     0.3855176E-17
  11     0.6595807E-17
   1     0.9389718E-17
   2     0.7459847E-17
myid/meshtag/norphan/nnodes/nobc=4 2 1150 85000 0
myid/meshtag/norphan/nnodes/nobc=7 2 1138 85000 0
Tioga exited successfully

```

**Adaptive Hole Map** (./run.sh 12):
```
 Generating           24  partitions
 Finished generating grid ..
 Writing tecplot compatible files..
 All done
 # tioga test on           12  processes
 # tioga test : finished reading grids
  [tioga::performConnectivity::getAdaptiveHoleMap] Mesh Body 0: Levels 8, total octants 34353, total leafs 30059
 connectivity time=   9.9077000000000026E-002
           1           7
           0           0
           0           0
           0           0
           0           0
           2          14
           0           0
           0           0
           3          21
           0           0
           0           0
           5          35
           0           0
           0           0
           2          14
           0           0
           0           0
           2          14
           0           0
           0           0
           0           0
        1202       10818
        1282       11538
           0           0
 data update time=   4.9900000000000500E-003
-- Interpolation error statistics --
ProcId    Error
   0     0.6619866E-17
   2     0.7459847E-17
   4     0.6485183E-17
   6     0.6526902E-17
   7     0.3255305E-16
   9     0.8691998E-17
   1     0.9389718E-17
   3     0.5463998E-17
   5     0.4214754E-17
   8     0.7567119E-17
  10     0.3855176E-17
  11     0.6595807E-17
myid/meshtag/norphan/nnodes/nobc=7 2 1055 85000 0
myid/meshtag/norphan/nnodes/nobc=4 2 1058 85000 0
Tioga exited successfully
```
![AdaptiveHoleMap sphere 8 levels](https://user-images.githubusercontent.com/37650539/213567727-6231955d-888b-4e74-92ce-49c3b3e4d21b.png)

ackirby88 committed:
Fixing header of meshblock output of tecplot files.

ackirby88 committed:
Fix cmakelists for driver and gridgen mpi and tioga library linking.

ackirby88 committed:
WIP: check in the adaptive hole map source code.

ackirby88 committed:
Implemented adaptive hole map and initial tests for tioga test sphere case with multiblock looks coorrect. Need to extensively test.

michaeljbrazell commented:
nice, way faster!

ackirby88 committed:
(1) Add duplicate node check when an element has two adjoining faces with an outer/overset bc face and wall bc face -- this prevents the adaptive hole map from refining to the max number of levels. (2) Fix compiler warnings about the intstring character array buffer size.

ackirby88 committed:
Replaced octant pointers with std::vectors. Verified correctness and memory cleanup. Note: suspect small memory leak in exchangeBoxes().

ackirby88 committed:
Reenable poisson exe by fixing the fortran MPI compling issue.

jrood-nrel commented:
Can there be something other than a define to turn this on? Also, can we verify this has been run through ASAN or valgrind to see that there are no leaks?

ackirby88 committed:
Add tioga interface function to set hole map algorithm (instead of a compiler define variable). Also added memory cleanup for tioga shutdown -- checked with valgrind (iblank_cell is lost but only because it is handled by nalu-wind).

ackirby88 commented:
Hi Jon,
I added a new tioga interface function to set the hole map algorithm:  
> ```void tioga_setholemapalg_(int *alg); // [0] original hole map, [1] adaptive hole map```

This function replaces the compile-time define variable.

Second, I added some memory cleanup for tioga instance deletion.
There were no issues previously for repeated hole map constructions, but only when ```tioga_delete``` was called.
I verifed no memory leaks using Valgrind (single rank and 2 mpi ranks  -- output files attached below).

[valgrid.adaptHoleMap.1rank.txt](https://github.com/Exawind/tioga/files/10503072/valgrid.adaptHoleMap.1rank.txt)
[valgrid.adaptHoleMap.2ranks.txt](https://github.com/Exawind/tioga/files/10503073/valgrid.adaptHoleMap.2ranks.txt)

ackirby88 committed:
Merge branch 'adaptive-hole-map' of https://github.com/ackirby88/tioga into adaptive-hole-map

ackirby88 committed:
Setting the hole map algorithm to the original for merging.

jrood-nrel commented on src/bookKeeping.C:
```diff
@@ -348,6 +349,198 @@ void MeshBlock::processDonors(HOLEMAP *holemap, int nmesh, int **donorRecords,do
+    verbose=0;
+    if(iblank[i] < 0){
+      temp=donorList[i];
+      while(temp!=NULL){
```
Does this need to exist?

ackirby88 replied:
It's a hold-over from the original function lines 320-330, so I wouldn't change the functionality.
But it does look potentially buggy with the while construct. Let me track down the history of the donorList.

ackirby88 replied:
I'm wondering if there's a bug in the original function. My guess is there should be
'''temp=temp->next;'''
Inside the while loop. Thoughts ?

psakievich replied:
it doesn't make any sense without some kind of increment inside the `while`

ashesh2512 replied:
Does indeed seem suspect ... Do you see a reason why it hasn't broken anything yet?

ackirby88 replied:
I think it worked before because Jay already sorted the temp donor resolutions (finest donor first) inside `insertInList` and already established the iblank value to be a fringe (-1). Otherwise, it would have a better resolution than the donor and not entered that if-statement check.

ackirby88 committed:
Revome old svn files.

ackirby88 committed:
(1) Fix while loop bug which didn't update the temp donorList candidate in MeshBlock::processDonors. (2) Replace raw c-pointers for adaptive processDonors.

ackirby88 committed:
Merge branch 'adaptive-hole-map' of https://github.com/ackirby88/tioga into adaptive-hole-map

ashesh2512 commented:
@ackirby88 Where is the `.svn` folder coming from?

ackirby88 commented:
> Where is the `.svn` folder coming from?

@ashesh2512 It was from the Jay's original TIOGA repository.

ashesh2512 commented on driver/CMakeLists.txt:
```diff
@@ -52,7 +53,7 @@ target_link_libraries(poisson_mms.exe tiogadriver)
 #endif()
 
-install(TARGETS tiogadriver tioga.exe tioga_read.exe poisson_mms.exe
+install(TARGETS tiogadriver tioga.exe tioga_read.exe #poisson_mms.exe
```
Does poisson_mms.exe still work?

ashesh2512 commented on driver/CMakeLists.txt:
```diff
@@ -31,8 +32,8 @@ target_link_libraries(tioga.exe tiogadriver)
 
-add_executable(poisson_mms.exe poisson_mms.f90)
-target_link_libraries(poisson_mms.exe tiogadriver)
+#add_executable(poisson_mms.exe poisson_mms.f90)
```
if poisson_mms.exe is obsolete, we should just remove it

ashesh2512 commented on src/codetypes.h:
```diff
@@ -55,8 +57,31 @@
 #define HOLEMAPSIZE        192
 // #define NFRINGE            3
 // #define NVAR               6
```
can we remove the commented out lines here?

ackirby88 replied:
They're legacy so your call.

ashesh2512 commented on src/codetypes.h:
```diff
@@ -55,8 +57,31 @@
 #define HOLEMAPSIZE        192
```
Can you elaborate what this does? I see it being used only here - `dsbox = dsmax / HOLEMAPSIZE;` How did we arrive at 192?

ackirby88 replied:
This number is the number of auxiliary mesh elements in each director of the uniform Cartesian hole map size (192 x 192 x 192). For every mesh body, a hole map is that size is constructed (with each cell marked as a hole or outer value). Originally, Jay had HOLEMAPSIZE as 64, which is actually too small because you would likely have nodes marked as wall-bc and outer-bc falling into the same Cartesian element (which is prevents the correct identification of hole cells).
The number 192 was set by your team (Shreyas) -- it's a nice compromise between taking tons of memory and the computational speed to traverse the hole Cartesian map (called a SAM-Structured Auxiliary Mesh). The most you could really go (depending on how many cores and RAM/core) is ~768 -- but the wall-clock time explodes along with the memory.

ashesh2512 commented on src/codetypes.h:
```diff
@@ -55,8 +57,31 @@
+                               // [1] face-box intersection (water-tight)
+
+/* Fixed Octree Constraints: Do Not Change */
+#define OCTANT_MAXLEVEL     19
```
why 19?

ackirby88 replied:
I need to check if we can move it up to 30 levels for 3D.
I'm following the p4est implementation. It has to relate to the number of bits required for storing 3D coordinates (x/y/z).

ashesh2512 commented on src/codetypes.h:
```diff
@@ -25,11 +25,13 @@
 /*  Floating point definition                                         */
 /*====================================================================*/
 # define REAL double
+typedef int32_t qcoord_t;
```
Using qcoord_t feels confusing.

ackirby88 replied:
Following p4est implementation -- stands for quadrant coordinate type.

ashesh2512 commented on src/codetypes.h:
```diff
@@ -55,8 +57,31 @@
+/** The length of a octant of level l */
+#define OCTANT_LEN(l) ((qcoord_t) 1 << (OCTANT_MAXLEVEL - (l)))
+/** Conversion from integer coordinates to double coordinates */
+#define INT2DBL ((double) 1.0 / (double) OCTANT_ROOT_LEN)
```
in the code, sometimes this MACRO is used, and sometimes int2dbl is computed locally. I will list out those places as I start reviewing those files.

ashesh2512 commented on src/codetypes.h:
```diff
@@ -76,9 +101,87 @@
+  uint32_t id;      /**< [4B] element id on level */
+  uint8_t filltype; /**< [1B] floodfill: [0] inside SB, [1] outside SB, [2] hole SB */
+  uint8_t refined;  /**< [1B] flag if refined (i.e. is a parent) */
+  struct octant_full *nhbr[6];     /**< [48B] neighbor octant list */
```
this is just the face neighbors, yes?

ackirby88 replied:
yes

ashesh2512 commented on src/MeshBlock.C:
```diff
@@ -128,6 +129,7 @@ void MeshBlock::preprocess(void)
   obb=(OBB *) malloc(sizeof(OBB));
   findOBB(x,obb->xc,obb->dxc,obb->vec,nnodes);
   tagBoundary();
+  tagBoundaryFaces();
```
how does this modify the existing pathway which doesn't use the octree search?

ashesh2512 commented on src/codetypes.h:
```diff
@@ -55,8 +57,31 @@
+/* ADAPTIVE HOLE MAP OCTANT INFO                                    */
+/*==================================================================*/
+//#define USE_ADAPTIVE_HOLEMAP 1 // [0] original hole map, [1] adaptive hole map
+#define INTERSECT_ALG        1 // [0] point-box inclusion only
```
What is the scenario for not doing a water tight approach?

ackirby88 replied:
In this picture, the point-box algorithm would only tag the cells with the end-point nodes in them, missing the cells with surface intersections (any cell that has a line through it). Here, the middle element and the middle-bottom violate the outer-surface, wall-surface principal. This will cause the hole-map construction to fail at forming a fully-closed hole map.
![Alg-Type](https://user-images.githubusercontent.com/37650539/236029158-3df6ce67-b473-416d-9b70-45fce91742f9.png)

ashesh2512 commented on src/codetypes.h:
```diff
@@ -55,8 +57,31 @@
-/* inline debugging tools                                             */
+/* ADAPTIVE HOLE MAP OCTANT INFO                                    */
+/*==================================================================*/
+//#define USE_ADAPTIVE_HOLEMAP 1 // [0] original hole map, [1] adaptive hole map
```
left over comment

ashesh2512 commented on src/MeshBlock.C:
```diff
@@ -826,7 +1002,265 @@ void MeshBlock::markWallBoundary(int *sam,int nx[3],double extents[6])
+                                       level_octant_t *level,
+                                       uint8_t *taggedList,
+                                       uint8_t *tagList){
+#if (INTERSECT_ALG==1)
```
The need for different algorithms is not apparent to me. Could you please elaborate?

ackirby88 replied:
See the other comment.

ashesh2512 commented on src/MeshBlock.h:
```diff
@@ -387,6 +427,34 @@ class MeshBlock
+    int id;
+    int nrank;
+    int masterID;   /* master rank for distributing mesh block data */
+    MPI_Comm comm;  /* communicator containing all complement ranks + master */
```
what is a complement rank?

ackirby88 replied:
The term complement refers to a mathematical definition (The complement of set A is defined as a set that contains the elements present in the universal set but not in set A -- see [Math Complement Figure](https://www.onlinemathlearning.com/image-files/complement-set.png)). 

I'm constructing a communicator containing all of the mpi ranks not a part of a mesh block and the rank-master of the mesh block (so the master can send the hole map after it is constructed to all of the ranks not involved in the mesh block).

ashesh2512 commented on src/MeshBlock.h:
```diff
@@ -387,6 +427,34 @@ class MeshBlock
+    int nrank;
+    int masterID;   /* master rank for distributing mesh block data */
+    MPI_Comm comm;  /* communicator containing all complement ranks + master */
+    MPI_Request req;
```
is `req` used anywhere?

ackirby88 replied:
I don't think so -- I originally had asynchronous communications.
I'll check to see if I use it for the abutting mesh implementation (next week); if not, I'll remove it.

ashesh2512 commented on src/tiogaInterface.C:
```diff
@@ -152,6 +152,7 @@ extern "C" {
 
   void tioga_preprocess_grids_(void)
   {
+    tg->assembleComms();
```
Is this needed for non-adaptive holemap too? How does this modify the default pathway?

ackirby88 replied:
We can put a conditional around it.

ashesh2512 commented on src/MeshBlock.C:
```diff
@@ -826,7 +1002,265 @@ void MeshBlock::markWallBoundary(int *sam,int nx[3],double extents[6])
+
+  const uint32_t noctants = level->elem_count;
+  const qcoord_t levelh = OCTANT_LEN(level->level_id); // integer length of octant
+  const double int2dbl = (double) 1.0 / (double) OCTANT_ROOT_LEN; // integer length conversion
```
We should use the macro here.

ashesh2512 commented on src/MeshBlock.C:
```diff
@@ -826,7 +1002,265 @@ void MeshBlock::markWallBoundary(int *sam,int nx[3],double extents[6])
+
+  const uint32_t noctants = level->elem_count;
+  const qcoord_t levelh = OCTANT_LEN(level->level_id); // integer length of octant
+  const double int2dbl = (double) 1.0 / (double) OCTANT_ROOT_LEN; // integer length conversion
```
We should use the macro here.

ackirby88 commented:
@ashesh2512 I don't see your commits here...

ashesh2512 commented:
> @ashesh2512 I don't see your commits here...

@ackirby88 Turns out I never submitted the comments. You should be able to see them now.

ackirby88 committed:
Address comments from Ashesh: (1) conditionals around adaptive hole map specific functions (comms), (2) macro replacement, (3) increase max level to all full 32-bit usage, (4) add back poisson_mms.

ackirby88 commented:
@ashesh2512 I committed your recommended changes. Please give the built-in test case a try along with your other tests.

jrood-nrel commented on src/tioga.h:
```diff
@@ -24,6 +24,7 @@
 #include <vector>
 #include <map>
 #include <memory>
+#include <malloc.h>
```
I think this header should be removed.

jrood-nrel replied:
As in it's not available for all compilers.

ackirby88 committed:
Remove malloc.h as it's nonstandard; commented out allocated data function using mallinfo.

ackirby88 committed:
Merging the std:min max from exawind.

ackirby88 committed:
Merge branch 'Exawind-exawind' into adaptive-hole-map

ackirby88 commented:
@jrood-nrel @ashesh2512 I've merged in the latest exawind changes with the std::min/max fixes. I've also addressed your suggestions, Jon, about the malloc.h.
This may finally be a good time from NREL's side to complete this pull request and merge in these changes so we stop diverging.

ashesh2512 commented:
Merged through #14
